gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/LaunchDagProc.java

```diff
@@ -0,0 +1,205 @@
+      nextJobNames.add(DagManagerUtils.getJobName(dagNode));
+    }
+
+    log.info("Submitting next nodes for dagId {}, where next jobs to be submitted are {}", dagId, nextJobNames);
+    //Checkpoint the dag state
+    newDagManager.getDagStateStore().writeCheckpoint(dag);
+
+    Map<String, Set<Dag.DagNode<JobExecutionPlan>>> dagIdToNextJobs = Maps.newHashMap();
+    dagIdToNextJobs.put(dagId, nextNodes);
+    return dagIdToNextJobs;
+  }
+
+  /**
+   * Submits a {@link JobSpec} to a {@link org.apache.gobblin.runtime.api.SpecExecutor}.
+   */
+  private void submitJob(DagManagementStateStore dagManagementStateStore, Dag.DagNode<JobExecutionPlan> dagNode) {
+    DagManagerUtils.incrementJobAttempt(dagNode);
+    JobExecutionPlan jobExecutionPlan = DagManagerUtils.getJobExecutionPlan(dagNode);
+    jobExecutionPlan.setExecutionStatus(ExecutionStatus.RUNNING);
+    JobSpec jobSpec = DagManagerUtils.getJobSpec(dagNode);
+    Map<String, String> jobMetadata = TimingEventUtils.getJobMetadata(Maps.newHashMap(), jobExecutionPlan);
+
+    String specExecutorUri = DagManagerUtils.getSpecExecutorUri(dagNode);
+
+    // Run this spec on selected executor
+    SpecProducer<Spec> producer;
+    try {
+      this.newDagManager.getUserQuotaManager().checkQuota(Collections.singleton(dagNode));
+      producer = DagManagerUtils.getSpecProducer(dagNode);
+      TimingEvent jobOrchestrationTimer = this.eventSubmitter.getTimingEvent(TimingEvent.LauncherTimings.JOB_ORCHESTRATED);
+
+      // Increment job count before submitting the job onto the spec producer, in case that throws an exception.
+      // By this point the quota is allocated, so it's imperative to increment as missing would introduce the potential to decrement below zero upon quota release.
+      // Quota release is guaranteed, despite failure, because exception handling within would mark the job FAILED.
+      // When the ensuing kafka message spurs DagManager processing, the quota is released and the counts decremented
+      // Ensure that we do not double increment for flows that are retried
+      if (dagNode.getValue().getCurrentAttempts() == 1) {
+        NewDagManager.getDagManagerMetrics().incrementRunningJobMetrics(dagNode);
+      }
+      // Submit the job to the SpecProducer, which in turn performs the actual job submission to the SpecExecutor instance.
+      // The SpecProducer implementations submit the job to the underlying executor and return when the submission is complete,
+      // either successfully or unsuccessfully. To catch any exceptions in the job submission, the DagManagerThread
+      // blocks (by calling Future#get()) until the submission is completed.
+      Future<?> addSpecFuture = producer.addSpec(jobSpec);
+      dagNode.getValue().setJobFuture(Optional.of(addSpecFuture));
+      //Persist the dag
+      newDagManager.getDagStateStore().writeCheckpoint(dagManagementStateStore.getDag(DagManagerUtils.generateDagId(dagNode).toString()));
```

Comment:
again, why not keep so much state floating around, rather than consolidate it into a singular `facade` we can easily replace--all-or-nothing?